src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2693 @@
+import Concordium.Types.Updates (UpdateInstruction (uiSignHash))
+import Control.Concurrent.Async
+import Control.Monad.Reader
+import Control.Monad.State.Strict
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.ByteString (ByteString)
+import Data.IORef
+import Data.Maybe (fromJust, maybeToList)
+import Data.ProtoLens (defMessage)
+import Data.ProtoLens.Field qualified as Field
+import Data.ProtoLens.Service.Types
+import Data.Sequence (Seq (Empty))
+import Data.String
+import Data.Time.Clock.POSIX (posixSecondsToUTCTime)
+import Proto.V2.Concordium.Service qualified as CS
+import Proto.V2.Concordium.Types qualified as ProtoFields
+import Proto.V2.Concordium.Types_Fields qualified as Proto
+
+{- |A helper function that serves as an inverse to `mkSerialize`,
+
+ Converts a protocol buffer wrapper type to a native Haskell type
+ which is a member of `Serialize`.
+
+ More concretely, the wrapper type should be of the form
+
+ > message Wrapper {
+ >    bytes value = 1
+ > }
+
+ where the name @Wrapper@ can be arbitrary, but the @value@ field must exist,
+ and it must have type @bytes@.
+-}
+deMkSerialize ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        BS.ByteString
+    , S.Serialize a
+    ) =>
+    b ->
+    Maybe a
+deMkSerialize val = case S.decode $ val ^. ProtoFields.value of
+    Left _ -> Nothing
+    Right v -> v
+
+-- |Like 'deMkSerialize' above, but for Word64 fields.
+deMkWord64 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word64
+    ) =>
+    b ->
+    Word64
+deMkWord64 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word32 fields instead.
+deMkWord32 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Word32
+deMkWord32 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word16 fields instead.
+deMkWord16 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word16
+deMkWord16 val =
+    if v <= 65_535
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+-- |Like 'mkWord32', but for Word8 fields instead.
+deMkWord8 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word8
+deMkWord8 val =
+    if v <= 255
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+{- |A helper class analogous to something like Aeson's FromJSON.
+ It exists to make it more manageable to convert the Protobuf types to
+ their internal Haskell type equivalents.
+-}
+class FromProto a where
+    -- |The corresponding Haskell type.
+    type Output' a
+
+    -- |A conversion function from the protobuf type to its Haskell equivalent.
+    fromProto :: a -> Maybe (Output' a)
+
+fromProtoError :: Maybe a
```

Comment:
I added this at some point because I thought it might make it easier to change the (conversion error) wrapper monad eventually, but it seems I forgot to use it.